kernel/arch/m68k/Hat.cpp

```diff
@@ -0,0 +1,203 @@
+#include <arch/Hat.h>
+#include <debug/Log.h>
+#include <memory/Pmm.h>
+#include <Memory.h>
+#include <Maths.h>
+
+#define PFLUSH(vaddr) do { asm volatile("pflush %0" :: "a"(vaddr) : "memory"); } while(false)
+#define PFLUSHA() do { asm volatile("pflushan" ::: "memory"); } while(false)
+
+namespace Npk
+{
+    constexpr size_t PageTableEntries = 128;
+    constexpr size_t PagingLevels = 3;
+    constexpr uint32_t ResidentFlag = 3;
+    constexpr uint32_t WriteProtectFlag = 1 << 2;
+    constexpr uint32_t GlobalFlag = 1 << 10;
+
+    struct PageTable
+    {
+        uint32_t ptes[PageTableEntries];
+    };
+
+    struct HatMap
+    {
+        PageTable* root;
+    };
+
+    struct WalkResult
+    {
+        size_t level;
+        uint32_t* pte;
+        bool complete;
+    };
+
+    HatMap kernelMap;
+    uint32_t tableAddrMask;
+    uint32_t descAddrMask;
+
+    constexpr HatLimits limits
+    {
+        .flushOnPermsUpgrade = false,
+        .modeCount = 1,
+        .modes = {{ .granularity = PageSize }}
+    };
+
+    static inline void GetIndices(uintptr_t vaddr, size_t* indices)
+    {
+        indices[3] = (vaddr >> 25) & 0x7F;
+        indices[2] = (vaddr >> 18) & 0x7F;
+        indices[1] = (vaddr >> 12) & 0x3F;
+    }
+
+    static inline WalkResult WalkTables(PageTable* root, uintptr_t vaddr)
+    {
+        size_t indices[PagingLevels + 1];
+        GetIndices(vaddr, indices);
+
+        WalkResult result {};
+        PageTable* pt = AddHhdm(root);
+        for (size_t i = PagingLevels; i > 0; i--)
+        {
+            result.pte = &pt->ptes[indices[i]];
+            result.level = i;
+            if ((*result.pte & ResidentFlag) == 0)
+            {
+                result.complete = false;
+                return result;
+            }
+
+            pt = reinterpret_cast<PageTable*>((*result.pte & descAddrMask) + hhdmBase);
+        }
+
+        result.complete = *result.pte & ResidentFlag;
+        return result;
+    }
+
+    static inline uint32_t AllocPageTable(uint32_t* pte, size_t index)
+    {
+        const size_t clusterStride = PageSize / 1;
```

Comment:
dividing by 1?

Reply:
a placeholder value, there is a bug in the code when the value is > 1